Civi/Api4/Action/Queue/RunLoop.php

```diff
@@ -0,0 +1,128 @@
+ * sysadmins who want to fine-tune runners on a specific queue.
+ *
+ * Basically, this runs until the queue is empty -- or until you reach one of the limits.
+ *
+ * For more functionality (queue auto-detection, fatal recovery, more tunable limits,
+ * process pooling, privilege separation, wait-support), you should look to `coworker` or
+ * write a custom main-loop.
+ *
+ * @method ?string getQueue
+ * @method $this setQueue(?string $queue)
+ * @method ?int getMaxRequests()
+ * @method $this setMaxRequests(?int $maxRequests)
+ * @method ?int getMaxDuration()
+ * @method $this setMaxDuration(?int $maxDuration)
+ */
+class RunLoop extends \Civi\Api4\Generic\AbstractAction {
+
+  /**
+   * Name of the target queue.
+   *
+   * @var string
+   * @required
+   */
+  protected $queue;
+
+  /**
+   * Maximum number of tasks to execute.
+   *
+   * After reaching this limit, the loop will stop taking new tasks.
+   *
+   * @var int|null
+   *   Special values `null` and `-1` indicate infinite execution.
+   * @see \Civi\Coworker\Configuration::$maxWorkerRequests
+   */
+  public $maxRequests = 10;
+
+  /**
+   * Maximum amount of time (seconds) for which a single worker should execute.
+   *
+   * After reaching this limit, no more tasks will be given to the worker.
+   *
+   * @var int|null
+   *   Special values `null` and `-1` indicate infinite execution.
+   * @see \Civi\Coworker\Configuration::$maxWorkerDuration
+   */
+  public $maxDuration = 120;
+
+  public function _run(Result $result) {
+    $queue = \Civi::queue($this->queue);
+    $startTime = microtime(TRUE);
+    $requests = 0;
+    $errors = 0;
+    $successes = 0;
+    $message = NULL;
+    $perm = $this->getCheckPermissions();
+
+    while (TRUE) {
+      if (!$queue->isActive()) {
+        break;
+      }
+
+      if (static::isFinite($this->maxRequests) && $requests >= $this->maxRequests) {
+        $message = sprintf('Reached request limit (%d)', $this->maxRequests);
+        break;
+      }
+      if (static::isFinite($this->maxDuration) && (microtime(TRUE) - $startTime) >= $this->maxDuration) {
+        $message = sprintf('Reached duration limit (%d)', $this->maxDuration);
+        break;
+      }
+
+      try {
+        $requests++;
+
+        $claims = Queue::claimItems($perm)->setQueue($this->queue)->execute()->getArrayCopy();
+        if (empty($claims)) {
+          $message = 'No claimable items';
+          break;
+        }
+
+        $batchResults = Queue::runItems($perm)->setQueue($this->queue)->setItems($claims)->execute();
+      }
+      catch (\Throwable $t) {
+        $errors++;
+        $message = sprintf('Queue-item raised unhandled exception (%s: %s)', get_class($t), $t->getMessage());
+        break;
+      }
+
+      foreach ($batchResults as $batchResult) {
+        if ($batchResult['outcome'] === 'ok') {
+          $successes++;
+        }
+        else {
+          $errors++;
+          // Should we stop? No, we're just reporting stats.
```

Comment:
So - reading this - we would continue to process items from a DIFFERENT queue but not the same one wouldn't we?

Reply:
@eileenmcnaughton Good q. It should break down like this

* In the call `cv api4 Queue.runLoop queue=my-stuff `, it's scoped to one specific queue. The `queue` parameter is required. 
* If that queue has `error=>abort` policy and then becomes `status=>aborted`, then `runLoop` stops.
* You could have two different queues running two copies of `Queue.runLoop` . Then yes -- the other queue continues executing.

(*Note to self: It should probably return a better message. All the other exit conditions have a descriptive message to facilitate debugging...*)

Reply:
so @totten `runLoop` is basically `runQueue` & `runAll` is `runAll` queues ?